File: LeetCode/python3/37. Sudoku Solver.py
from typing import List


class Solution:
    def is_valid(self, board: list[list[str]], num: str, row: int, col: int) -> bool:
        for i in range(9):
            #  check each value in the row
            if num == board[row][i] and i != col:
                return False

            #  check each value in the column
            if num == board[i][col] and i != row:
                return False

        #  calculate which box
        box_row = row // 3
        box_col = col // 3

        # check each value in the box
        for i in range(box_row * 3, box_row * 3 + 3):
            for j in range(box_col * 3, box_col * 3 + 3):
                if i == row and j == col:
                    continue

                if board[i][j] == num:
                    return False

        return True

    def solveSudoku(self, board: List[List[str]]) -> None:
        possible = tuple(
            ((row, col), tuple(num for num in map(str, range(1, 10)) if self.is_valid(board, num, row, col)))
            for row in range(9) for col in range(9) if board[row][col] == '.'
        )

        stack = [[0, 0]]

        while stack:
            index, num_index = stack[-1]
            if index == len(possible):
                break

            (row, col), nums = possible[index]

            if num_index >= len(nums):
                board[row][col] = '.'
                stack.pop()
                continue

            stack[-1][1] += 1

            if self.is_valid(board, nums[num_index], row, col):
                board[row][col] = nums[num_index]
                stack.append([index + 1, 0])

        print(board)


if __name__ == "__main__":
    print(Solution().solveSudoku(
        [
            ["5", "3", ".", ".", "7", ".", ".", ".", "."],
            ["6", ".", ".", "1", "9", "5", ".", ".", "."],
            [".", "9", "8", ".", ".", ".", ".", "6", "."],
            ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
            ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
            ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
            [".", "6", ".", ".", ".", ".", "2", "8", "."],
            [".", ".", ".", "4", "1", "9", ".", ".", "5"],
            [".", ".", ".", ".", "8", ".", ".", "7", "9"]
        ]))

File: LeetCode/rust/37. Sudoku Solver.rs
struct Solution;

impl Solution {
    fn is_valid(board: &[Vec<char>], num: char, row: usize, col: usize) -> bool {
        if (0..9).any(|i| num == board[row][i] && i != col || num == board[i][col] && i != row) {
            return false;
        }

        let box_row_start: usize = row / 3usize * 3usize;
        let box_col_start: usize = col / 3usize * 3usize;

        !(box_row_start..box_row_start + 3).any(|i| {
            (box_col_start..box_col_start + 3).any(|j| {
                (i, j) != (row, col) && board[i][j] == num
            })
        })
    }

    pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
        let possible: Vec<([usize; 2], Vec<char>)> = (0..9)
            .flat_map(|row| (0..9).map(move |col| (row, col)))
            .filter(|&(row, col)| board[row][col] == '.')
            .map(|(row, col)| {
                (
                    [row, col],
                    ('1'..='9')
                        .filter(|&num| Self::is_valid(board, num, row, col))
                        .collect::<Vec<char>>()
                )
            })
            .collect::<Vec<([usize; 2], Vec<char>)>>();

        let mut stack: Vec<[usize; 2]> = vec![[0, 0]];

        while !stack.is_empty() {
            let [index, num_index] = *stack.last().unwrap();

            if index == possible.len() {
                break;
            }

            let ([row, col], nums) = (possible[index].0, &possible[index].1);

            if num_index >= nums.len() {
                board[row][col] = '.';
                stack.pop();
                continue;
            }

            stack.last_mut().unwrap()[1] += 1;

            if Self::is_valid(board, nums[num_index], row, col) {
                board[row][col] = nums[num_index];
                stack.push([index + 1usize, 0usize]);
            }
        }
    }
}

fn main() {
    let mut board: Vec<Vec<char>> = vec![
        vec!['5', '3', '.', '.', '7', '.', '.', '.', '.'],
        vec!['6', '.', '.', '1', '9', '5', '.', '.', '.'],
        vec!['.', '9', '8', '.', '.', '.', '.', '6', '.'],
        vec!['8', '.', '.', '.', '6', '.', '.', '.', '3'],
        vec!['4', '.', '.', '8', '.', '3', '.', '.', '1'],
        vec!['7', '.', '.', '.', '2', '.', '.', '.', '6'],
        vec!['.', '6', '.', '.', '.', '.', '2', '8', '.'],
        vec!['.', '.', '.', '4', '1', '9', '.', '.', '5'],
        vec!['.', '.', '.', '.', '8', '.', '.', '7', '9'],
    ];

    Solution::solve_sudoku(&mut board);
    println!("{:?}", board)
}